File: s3_15654_NandM5/main.cpp
#include <iostream>
#include <vector>
#include <bitset>
#include <algorithm>

int N, M;
std::vector<int> arr;
std::bitset<9> visited;
std::vector<int> used;

void dfs(int k){
    if(k==M){
        for(int i=0; i<M; i++){
            std::cout << used[i] << " ";
        }
        std::cout << "\n";
        return;
    }
    for(int i=0; i<N; i++){
        if(visited[i]) continue;
        visited[i] = true;
        used[k] = arr[i];
        dfs(k+1);
        visited[i] = false;
    }
}


int main(){
    std::ios::sync_with_stdio(false);
    std::cin.tie(NULL); std::cout.tie(NULL);
    std::cin >> N >> M;
    arr = std::vector<int>(N);
    used = std::vector<int>(M);
    for(int i=0; i<N; i++){
        std::cin >> arr[i];
    }
    std::sort(arr.begin(), arr.end());
    dfs(0);

}

File: s3_15654_NandM5/src/main.rs
use std::io::{self, BufRead, Write};

fn dfs(k: usize, n: usize, m: usize, arr: &Vec<i32>, visited: &mut Vec<bool>, used: &mut Vec<i32>) {
    if k == m {
        for i in 0..m {
            print!("{} ", used[i]);
        }
        println!();
        io::stdout().flush().unwrap();
        return;
    }
    for i in 0..n {
        if visited[i] {
            continue;
        }
        visited[i] = true;
        used[k] = arr[i];
        dfs(k + 1, n, m, arr, visited, used);
        visited[i] = false;
    }
}

fn main() {
    let stdin = io::stdin();
    let mut iter = stdin.lock().lines().map(|l| l.unwrap())
        .flat_map(|line| line.split_whitespace().map(|num| num.parse::<i32>().unwrap()).collect::<Vec<i32>>().into_iter());
    
    let n = iter.next().unwrap() as usize;
    let m = iter.next().unwrap() as usize;
    
    let mut arr = vec![0; n];
    for i in 0..n {
        arr[i] = iter.next().unwrap();
    }
    arr.sort();
    
    let mut visited = vec![false; n];
    let mut used = vec![0; m];
    
    dfs(0, n, m, &arr, &mut visited, &mut used);
}